stdlib/src/builtin/string.mojo

```diff
@@ -1405,36 +1441,172 @@ struct String(
+            A List of Strings containing the input split by the separator.
 
-        Raises:
-          Error if an empty delimiter is specified.
-        """
-        if not delimiter:
-            raise Error("empty delimiter not allowed to be passed to split.")
+        Examples:
 
+        ```mojo
+        # Splitting a space
+        _ = String("hello world").split(" ") # ["hello", "world"]
+        # Splitting adjacent separators
+        _ = String("hello,,world").split(",") # ["hello", "", "world"]
+        # Splitting with maxsplit
+        _ = String("1,2,3").split(",", 1) # ['1', '2,3']
+        # Splitting an empty string or filled with whitespaces
+        _ = String("      ").split() # []
+        _ = String("").split() # []
+        # Splitting a string with leading, trailing, and middle whitespaces
+        _ = String("      hello    world     ").split() # ["hello", "world"]
+        # Splitting a string full of the separator
+        _ = String(",,,").split(",") # ["", "", "", ""]
+        # Splitting adjacent universal newlines:
+        # _ = String("hello \\t\\n\\r\\f\\v\\x1c\\x1e\\x85world").split() # ["hello", "world"]
+        ```
+        .
+        """
         var output = List[String]()
 
-        var current_offset = 0
-        while True:
-            var loc = self.find(delimiter, current_offset)
-            # delimiter not found, so add the search slice from where we're currently at
-            if loc == -1:
-                output.append(self[current_offset:])
-                break
+        var str_len = len(self)
+        var lhs = 0
+        var rhs = 0
+        var items = 0
+        while lhs < str_len:
+            if separator == "":
+                # Python adds all "whitespace chars" as one separator
+                # if no separator was specified
+                while lhs < str_len:
+                    if not self[lhs] in String.WHITESPACE:
+                        break
+                    lhs += 1
+                # if it went until the end of the String, then
+                # it should be sliced up until the original
+                # start of the whitespace which was already appended
+                if lhs == str_len:
+                    break
+                rhs = lhs + 1
+                while rhs < str_len:
+                    if self[rhs] in String.WHITESPACE:
+                        break
+                    rhs += 1
+            else:
+                rhs = self.find(separator, lhs)
+                if rhs == -1:
+                    output.append(self[lhs:])
+                    break
+
+            if maxsplit > -1:
+                if items == maxsplit:
+                    output.append(self[lhs:])
+                    break
+                items += 1
+
+            output.append(self[lhs:rhs])
+            lhs = rhs + len(separator)
+
+        return output
+
+    fn split[separator: String = "", maxsplit: Int = -1](self) -> List[String]:
```

Comment:
Can you please remove this overload? It's not clear at this point if such optimization would pull it's weight.